adam_core/dynamics/ephemeris.py

```diff
@@ -0,0 +1,287 @@
+from ..orbits.ephemeris import Ephemeris
+from ..orbits.orbits import Orbits
+from .aberrations import _add_light_time, add_stellar_aberration
+
+MU = c.MU
+
+
+@jit
+def _generate_ephemeris_2body(
+    propagated_orbit: np.ndarray,
+    observation_time: float,
+    observer_coordinates: jnp.ndarray,
+    lt_tol: float = 1e-10,
+    mu: float = MU,
+    max_iter: int = 100,
+    tol: float = 1e-15,
+    stellar_aberration: bool = False,
+) -> Tuple[jnp.ndarray, jnp.float64]:
+    """
+    Given a propagated orbit, generate its on-sky ephemeris as viewed from the observer.
+    This function calculates the light time delay between the propagated orbit and the observer,
+    and then propagates the orbit backward by that amount to when the light from object was actually
+    emitted towards the observer ("astrometric coordinates").
+
+    The motion of the observer in an inertial frame will cause an object
+    to appear in a different location than its true location, this is known as
+    stellar aberration (often referred to in combination with other aberrations as "apparent
+    coordinates"). Stellar aberration can optionally be applied after
+    light time correction has been added but it should not be necessary when comparing to ephemerides
+    of solar system small bodies extracted from astrometric catalogs. The stars to which the
+    catalog is calibrated undergo the same aberration as the moving objects as seen from the observer.
+
+    If stellar aberration is applied then the velocity of the input orbits are unmodified, only the position
+    vector is modified with stellar aberration.
+
+    For more details on aberrations see:
+        https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/FORTRAN/req/abcorr.html
+        https://ssd.jpl.nasa.gov/horizons/manual.html#defs
+
+    Parameters
+    ----------
+    propagated_orbit : `~jax.numpy.ndarray` (6)
+        Barycentric Cartesian orbit propagated to the given time.
+    observation_time : float
+        Epoch at which orbit and observer coordinates are defined.
+    observer_coordinates : `~jax.numpy.ndarray` (3)
+        Barycentric Cartesian observer coordinates.
+    lt_tol : float, optional
+        Calculate aberration to within this value in time (units of days).
+    mu : float, optional
+        Gravitational parameter (GM) of the attracting body in units of
+        AU**3 / d**2.
+    max_iter : int, optional
+        Maximum number of iterations over which to converge for propagation.
+    tol : float, optional
+        Numerical tolerance to which to compute universal anomaly during propagation using the Newtown-Raphson
+        method.
+    stellar_aberration : bool, optional
+        Apply stellar aberration to the ephemerides.
+
+    Returns
+    -------
+    ephemeris_spherical : `~jax.numpy.ndarray` (6)
+        Topocentric Spherical ephemeris.
+    lt : float
+        Light time correction (t0 - corrected_t0).
+    """
+    # Add light time correction
```

Comment:
I guess it is the opposite? I think of it as correcting the observed position to take into account light travel time. But yes, your understanding is completely correct. We take the geometric position (light has infinite speed) and back-propagate the state vector to take into account that light actually has finite speed. The ground truth here is always the observation time so that is where we first propagate things to. Once you do that you have an estimate of the topocentric distance and then you iterate the state backwards to solve for the light travel time.